### PR TITLE
center O_k

### PR DESCRIPTION
**Pull request by inailuig**
Currently the on_the_fly SR code is using S = ⟨Ô†ΔÔ⟩ for the quantum geometric tensor.

This adds to option to use S = ⟨ΔÔ†ΔÔ⟩  with a centered wavefunction instead and makes it the default.

This might be better numerically, but comes at the cost of slightly more communication (one allreduce of the parameters) and one more forward pass with the slightly more complicated centered wavefunction.

### CODE REVIEW

File: netket/optimizer/sr/sr_onthefly_logic.py [hunk 1]
```diff
@@ -161,7 +165,18 @@ def Odagger_DeltaO_v(samples, params, v, forward_fn, vjp_fun=None):
     # v_tilde /= n_samples (elementwise):
     v_tilde = v_tilde * (1.0 / (samples.shape[0] * n_nodes))
 
-    return OH_w(samples, params, v_tilde, forward_fn, dtype=dtype, vjp_fun=vjp_fun)
+    # ⟨Ô^†ΔÔ⟩v
+    res = OH_w(samples, params, v_tilde, forward_fn, dtype=dtype, vjp_fun=vjp_fun)
+
+    if centered:
+        # ⟨ΔÔ⟩v; is a scalar close to 0; (MPI sum):
+        v_tilde_sum = _sum(v_tilde)
```
Comment (gcarleo) [inline netket/optimizer/sr/sr_onthefly_logic.py:173]: sorry it's not clear to me why you need to center it here explicitly, don't you already have a centered quantity when you compute the first product?

Comment (inailuig) [reply netket/optimizer/sr/sr_onthefly_logic.py:173]: Yes but only for O_l, not for O_k
I am not aware of a simple way to do `(jacobian + matrix) @ vector` as a jvp directly with jax (where the matrix is ⟨Ô⟩ broadcast to all columns)

File: netket/optimizer/sr/sr_onthefly_logic.py [hunk 2]
```diff
@@ -161,7 +165,18 @@ def Odagger_DeltaO_v(samples, params, v, forward_fn, vjp_fun=None):
...
+
+        # ⟨Ô^†ΔÔ⟩v - ⟨Ô⟩^†⟨ΔÔ⟩v = ⟨ΔÔ^†ΔÔ⟩v
+        # return res - omean.conj * v_tilde_sum:
+        return tree_axpy(-v_tilde_sum, tree_conj(omean), res)
```
Comment (inailuig) [inline netket/optimizer/sr/sr_onthefly_logic.py:177]:
```suggestion
        return tree_axpy(-v_tilde_sum, tree_conj(jax.tree_map(_sum_inplace, omean)), res)
```

**Comment by PhilipVinc** [comment]
@inailuig what's the status?

**Comment by inailuig** [comment]
Works and is ready for testing, but still needs some polishing if we decide to merge it.

File: netket/optimizer/sr/base.py [hunk 3]
```diff
@@ -31,3 +31,4 @@ class SR:
     """
 
     diag_shift: float = 0.01
+    centered: bool = False
```
[the remark this answers is not in the document]
Comment (PhilipVinc) [reply netket/optimizer/sr/base.py:34]: Also, add a docstring to the field mentioning that it computes <O\deltaO> instead of <deltaOdeltaO>.
Otherwise one might think centered=False is computing <OO>

File: netket/optimizer/sr/sr_onthefly_logic.py [hunk 4]
```diff
@@ -184,7 +223,14 @@ def mat_vec(v, forward_fn, params, samples, diag_shift):
...
 
-    res = Odagger_DeltaO_v(samples, params, v, forward_fn)
+    # TODO ideally we would make centered a static argument
+    # however fishing it out of the sr dataclass at jit is ugly
+    res = jax.lax.cond(
+        centered,
+        lambda args: DeltaOdagger_DeltaO_v(*args, forward_fn),
+        lambda args: Odagger_DeltaO_v(*args, forward_fn),
+        (samples, params, v),
+    )
```
Comment (PhilipVinc) [inline netket/optimizer/sr/sr_onthefly_logic.py:233]: centered shouldn't be a conditional.
let's assume it's a static flag.
```struct.field(pytree_node=False)```

**Review by PhilipVinc — CHANGES_REQUESTED** [review]

**Comment by PhilipVinc** [comment]
Also needs to edit `sr/api.py` to include the new flag

**Comment by PhilipVinc** [comment]
We will be merging so let's polish

File: netket/optimizer/sr/sr_onthefly_logic.py (hunk 4 above)
Comment (inailuig) [reply netket/optimizer/sr/sr_onthefly_logic.py]: > Also needs to edit `sr/api.py` to include the new flag

can't it just go into kwargs?

**Comment by inailuig** [comment]
> Also needs to edit `sr/api.py` to include the new flag

can't it just go into kwargs?

**Comment by PhilipVinc** [comment]
ah, yes.
I need to add docs to that though...

**Review by PhilipVinc — APPROVED** [review]

**Comment by PhilipVinc** [comment]
@inailuig it's done now?
Can I just ask you to run a quick benchmark and check how much the performance degrades on Examples/Ising1d/ising1d_sr.py for N=10,20,30..?

**Comment by inailuig** [comment]
I still have to finish improving the test a bit

**Comment by inailuig** [comment]
For alpha=1, n_samples=1000 there is no significant difference in performance on a single core:

not centered:
```
L 
10 59.16it/s, Energy=-12.7867 ± 0.0020
20 19.80it/s, Energy=-25.4919 ± 0.0040
40  6.50it/s, Energy=-50.9238 ± 0.0070
80  1.73it/s, Energy=-101.179 ± 0.057
```
centered:
```
L
10 59.59it/s, Energy=-12.7871 ± 0.0020
20 19.59it/s, Energy=-25.4940 ± 0.0056
40  6.38it/s, Energy=-50.9355 ± 0.0097
80  1.76it/s, Energy=-101.558 ± 0.050
```
---

**Comment by gcarleo** [comment]
Maybe a coincidence but 4/4 energies are lower in the case of centered O_k?

**Comment by gcarleo** [comment]
what does it happen if you use the same seed in both cases?

**Comment by inailuig** [comment]
I was using the same seed

**Comment by PhilipVinc** [comment]
that is good.
Worth setting it on by default @gcarleo ?

**Comment by gcarleo** [comment]
> I was using the same seed

Oh really? then it is absolutely remarkable that the energy difference is so large... are we sure the two versions are doing exactly the same thing (in the absence of finite precision issues) ?

**Comment by gcarleo** [comment]
> that is good.
> Worth setting it on by default @gcarleo ?

Yes I am now worried that the non-centered version is really very unstable...

File: netket/optimizer/sr/sr_onthefly.py [hunk 5]
```diff
@@ -41,6 +41,7 @@ class SR_otf(SR):
     atol: float = 0.0
     maxiter: int = None
     M: Optional[Union[Callable, Ndarray]] = None
+    centered: bool = struct.field(pytree_node=False, default=False)
```
Comment (PhilipVinc) [inline netket/optimizer/sr/sr_onthefly.py:44]:
```suggestion
    centered: bool = struct.field(pytree_node=False, default=True)
```

**Review by PhilipVinc — CHANGES_REQUESTED** [review]

**Comment by PhilipVinc** [comment]
@inailuig ping me when the tests are good.

**Comment by inailuig** [comment]
same but w/ 8192 samples:

non-centered:
```
L
10 8.10it/s, Energy=-12.78433 ± 0.00060
20 2.47it/s, Energy=-25.4899 ± 0.0013
```
centered:
```
L
10 7.66it/s, Energy=-12.78433 ± 0.00061
20 2.51it/s, Energy=-25.4896 ± 0.0012
```

File: netket/optimizer/sr/sr_onthefly_logic.py [hunk 6]
```diff
@@ -164,16 +164,58 @@ def Odagger_DeltaO_v(samples, params, v, forward_fn, vjp_fun=None):
...
+        params,
+        forward_fn,
+        dtype=dtype,
+        return_vjp_fun=False,
+        vjp_fun=vjp_fun,
+        allreduce=True,
+    )
+
+    def forward_fn_centered(params, x):
+        return forward_fn(params, x) - tree_dot(params, omean)
```
Comment (inailuig) [inline netket/optimizer/sr/sr_onthefly_logic.py:201]: It would be possible to use the same trick also in the non-centered version, 
that would make it almost as expensive as this though...

**Comment by inailuig** [comment]
> @inailuig ping me when the tests are good.
@PhilipVinc 
Just the same single test like in nk3 is failing.
